### PR TITLE
stake-pool: Add "IncreaseAdditionalValidatorStake" instruction

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

As brought up during the single validator manager proposal #3881, there's an annoying "attack" vector where a person puts in the minimum stake delegation at the start of an epoch and immediately increases stake on the validator, preventing later SOL deposits from being activated.

#### Solution

There's nothing technically stopping two activating stakes from merging, so introduce a new instruction on the stake pool program which allows for a second increase in an epoch.

The normal instruction splits into the transient account and then activates it, so this instruction:

* splits into an ephemeral stake account
* if the transient account exists: merges the ephemeral stake account into the transient stake account\
* if the transient account doesn't exist: splits from ephemeral into transient, so pools can transition to always using this new instruction

### CODE REVIEW

File: stake-pool/program/src/processor.rs [hunk 1]
```diff
@@ -1327,7 +1355,19 @@ impl Processor {
         }
         let mut validator_stake_info = maybe_validator_stake_info.unwrap();
         if validator_stake_info.transient_stake_lamports > 0 {
-            return Err(StakePoolError::TransientAccountInUse.into());
+            if maybe_ephemeral_stake_seed.is_none() {
+                msg!("Attempting to increase stake on a validator with transient stake, use IncreaseAdditionalValidatorStake with the existing seed");
```
Comment (t-nelson) [inline stake-pool/program/src/processor.rs:1359]:
```suggestion
                msg!("Attempting to increase stake on a validator with pending transient stake, use IncreaseAdditionalValidatorStake with the existing seed");
```

File: stake-pool/program/src/processor.rs [hunk 2]
```diff
@@ -1410,37 +1435,136 @@ impl Processor {
...
         }
 
-        create_stake_account(
-            transient_stake_account_info.clone(),
-            transient_stake_account_signer_seeds,
-            system_program_info.clone(),
-        )?;
+        let maybe_split_from_account_info =
+            if let Some(ephemeral_stake_seed) = maybe_ephemeral_stake_seed {
+                let ephemeral_stake_account_info = maybe_ephemeral_stake_account_info.unwrap();
```
Comment (t-nelson) [inline stake-pool/program/src/processor.rs:1440]: consider `zip`ing the `Option`, either in the preceding `match` or above, where these are resolved. no sense getting the auditors' hopes up with an `unwrap`!

File: stake-pool/program/src/processor.rs [hunk 3]
```diff
@@ -1410,37 +1435,136 @@ impl Processor {
...
+                    validator_vote_account_info.clone(),
+                    clock_info.clone(),
+                    stake_history_info.clone(),
+                    stake_config_info.clone(),
+                    withdraw_authority_info.clone(),
+                    stake_pool_info.key,
+                    AUTHORITY_WITHDRAW,
+                    stake_pool.stake_withdraw_bump_seed,
+                )?;
+                if validator_stake_info.transient_stake_lamports > 0 {
```
Comment (t-nelson) [inline stake-pool/program/src/processor.rs:1482]: could use this condition to select the split destination.  would dereplicode a bit

**Review by t-nelson — DISMISSED** [review]
just a couple nits here. lgtm!

File: stake-pool/program/src/processor.rs [hunk 4]
```diff
@@ -1410,37 +1435,134 @@ impl Processor {
...
+            )?;
+        } else {
+            // no transient stake, split
+            let transient_stake_bump_seed = check_transient_stake_address(
+                program_id,
+                stake_pool_info.key,
+                transient_stake_account_info.key,
+                vote_account_address,
+                transient_stake_seed,
+            )?;
```
Comment (2501babe) [inline stake-pool/program/src/processor.rs:1511]: we should move the call to `check_transient_stake_address` outside the if/else. as it stands, if `validator_stake_info.transient_stake_lamports > 0` is true, `transient_stake_account_info` is never validated. the pool can be griefed by creating an unrelated stake account with the same lockup/authority to merge into instead, resulting in a miscalculation in `validator_stake_info.transient_stake_lamports` at the end of this function, which could have unknown downstream effects

also for the record today i learned that merging stake accounts requires them to have the same authority, the missing check baited me into partway writing a poc for loss of funds before i learned this fact :sweat_smile:

**Comment by 2501babe** [comment]
other than that, looks good!

File: stake-pool/program/src/processor.rs (hunk 4 above)
Comment (joncinque) [reply stake-pool/program/src/processor.rs:1511]: Great catch, thanks! I don't think you could steal funds, but you could definitely mess up the pool management if you spoof this.

**Pull request merged**